### PR TITLE
Temporal Memory Examples (Python)

### PR DESCRIPTION
**Pull request by ctrl-z-9000-times**
I am considering dropping the following files since they are unit tests, not examples.  I can add them in a latter PR.
- py/src/nupic/examples/tm/tm_constant_test.py
- py/src/nupic/examples/tm/tm_overlapping_sequences.py
- py/src/nupic/examples/tm/tm_segment_learning.py
- py/src/nupic/examples/tm/tm_test.py

### CODE REVIEW

File: py/src/nupic/examples/tm/tm_high_order.py [hunk 1]
```diff
@@ -0,0 +1,307 @@
...
+# You should have received a copy of the GNU Affero Public License
+# along with this program.  If not, see http://www.gnu.org/licenses.
+#
+# http://numenta.org/licenses/
+# ----------------------------------------------------------------------
+
+__doc__ = """
+A tutorial that shows some features of the Temporal Memory.
+
+This program has demonstrates some basic properties of the
```
Comment (dkeeney) [inline py/src/nupic/examples/tm/tm_high_order.py]: An extra `has`.

Comment (dkeeney) [reply py/src/nupic/examples/tm/tm_high_order.py]: This is a very nice demonstration of TM. 👍

File: py/src/nupic/examples/tm/tm_overlapping_sequences.py [hunk 2]
```diff
@@ -0,0 +1,917 @@
...
+  # Keep these fixed:
+  connectedPerm = 0.5
+
+  tms = dict()
+
+  if includeCPP:
+    if VERBOSITY >= 2:
+      print("Creating BacktrackingTMCPP instance")
+
+    cpp_tm = BacktrackingTMCPP(numberOfCols = numCols, cellsPerColumn = cellsPerCol,
```
Comment (dkeeney) [inline py/src/nupic/examples/tm/tm_overlapping_sequences.py]: Again, the C++ part of BacktrackingTMCPP is gone.

File: py/src/nupic/examples/tm/tm_overlapping_sequences.py [hunk 3]
```diff
@@ -0,0 +1,917 @@
...
+
+
+  return True
+
+
+class TMOverlappingSeqsTest(testcasebase.TestCaseBase):
+
+  def testFastLearning(self):
+    """
+    Test with fast learning, make sure PAM allows us to train with fewer
```
Comment (dkeeney) [inline py/src/nupic/examples/tm/tm_overlapping_sequences.py]: What is PAM?   I assume P is probably `Prediction` but I don't know what 'A' and 'M' are.

File: py/src/nupic/examples/tm/tm_segment_learning.py [hunk 4]
```diff
@@ -0,0 +1,611 @@
...
+      maxSynapsesPerSegment = -1
+      globalDecay = 0.0001
+      maxAge = 1
+
+
+    if g_testCPPTM:
+      if g_options.verbosity > 1:
+        print("Creating BacktrackingTMCPP instance")
+
+      cppTM = BacktrackingTMCPP(numberOfCols = numCols, cellsPerColumn = 4,
```
Comment (dkeeney) [inline py/src/nupic/examples/tm/tm_segment_learning.py]: again...there is no C++ backtracking DM

**Comment by ctrl-z-9000-times** [comment]
For issue #216.

File: py/src/nupic/examples/tm/tm_overlapping_sequences.py (hunk 3 above)
Comment (ctrl-z-9000-times) [reply py/src/nupic/examples/tm/tm_overlapping_sequences.py]: PAM stands for "Pay Attention Mode". It was part of the backtracking TM.

Comment (dkeeney) [reply py/src/nupic/examples/tm/tm_overlapping_sequences.py]: Ah.  I guess I should have remembered that.

File: py/src/nupic/examples/tm/tm_segment_learning.py (hunk 4 above)
Comment (dkeeney) [reply py/src/nupic/examples/tm/tm_segment_learning.py]: ...I mean 'TM'

**Comment by ctrl-z-9000-times** [comment]
On further review, I'm definitely going to drop the following files from this PR:
-    py/src/nupic/examples/tm/tm_constant_test.py
-    py/src/nupic/examples/tm/tm_overlapping_sequences.py
-    py/src/nupic/examples/tm/tm_segment_learning.py
-    py/src/nupic/examples/tm/tm_test.py

Because:
- They are unit tests, outside of the scope of this PR.
- They don't have any print statements to guide the user through whats going on.  They don't even have very many inline comments.
- They all use the BacktrackingTM, not the normal TM.
- They are big, messy, and complicated so turning these files into presentable examples looks to be time consuming.

File: py/src/nupic/examples/tm/tm_high_order.py (hunk 1 above)
Comment (ctrl-z-9000-times) [reply py/src/nupic/examples/tm/tm_high_order.py]: Good spot!

File: py/src/nupic/examples/tm/tm_overlapping_sequences.py (hunk 2 above)
Comment (ctrl-z-9000-times) [reply py/src/nupic/examples/tm/tm_overlapping_sequences.py]: Removed this file.

File: bindings/py/cpp_src/bindings/algorithms/py_TemporalMemory.cpp [hunk 5]
```diff
@@ -316,6 +325,10 @@ R"(Returns the number of cells in this TemporalMemory.)");
         py_HTM.def("numberOfColumns", &HTM_t::numberOfColumns,
 R"(Returns the total number of mini-columns.)");
 
+        py_HTM.def_property_readonly("connections", [](const HTM_t &self)
+            { return self.connections; },
+R"(TODO: Document the dangers of using this...)");
```
Comment (ctrl-z-9000-times) [inline bindings/py/cpp_src/bindings/algorithms/py_TemporalMemory.cpp]: TODO before merging

**Comment by ctrl-z-9000-times** [comment]
Thanks for reviewing this!

File: py/src/nupic/examples/tm/hello_tm.py [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+    print("")
+
+
+print("################################################################################")
+print(__doc__)
+print("################################################################################")
+print("")
+print("Creating the Temporal Memory")
+tm = TM(columnDimensions = (50,),
+        cellsPerColumn=1,
```
Comment (breznak) [inline py/src/nupic/examples/tm/hello_tm.py]: TM with 1 cell/col does not really make sense (no higher order learning), unless it's necessary for the example, this should use "normal" number of cells, ie 4,8,...

File: py/src/nupic/examples/tm/tm_high_order.py [hunk 7]
```diff
@@ -0,0 +1,307 @@
...
+from nupic.algorithms import TemporalMemory as TM
+
+print("--------------------------------------------------")
+print(__doc__)
+print("--------------------------------------------------")
+print("")
+print("Creating the Temporal Memory")
+tm = TM(
+  columnDimensions = (2048,),
+  cellsPerColumn=8,
```
Comment (breznak) [inline py/src/nupic/examples/tm/tm_high_order.py]: ..oh, ok

**Comment by ctrl-z-9000-times** [comment]
> turned to unit-tests, or ran as a part of testing?

Not easily.  The primary output of these examples are the text they write and the figures they create.  IIRC We already have unit tests which do very similar things as these examples.

**Comment by breznak** [comment]
> We already have unit tests which do very similar things as these examples. 

ok, that was my concern, then cease closed :+1:

**Pull request merged**